docs/source/dev/py_postprocessing.rst

```diff
@@ -0,0 +1,178 @@
+widget logic and implements widgets for scan, simulation and iteration selection.
+
+When created with an ``experiment_path`` argument, it assumes that there will be a number of subdirectories containing
+scan results and named ``scan_<nr>``. Within each such directory it further assumes one or many PIConGPU simulation output directories
+named ``sim_<nr>``.
+The base class provides functionality to easily allow switching between visualizations for different scans or simulations.
+
+When implementing a new widget you have to perform the following steps:
+
+1. Let the widget class inherit from the base visualizer
+
+.. code:: python
+
+   from .base_visualizer import Visualizer as BaseWidgetVisualizer
+
+   class Visualizer(BaseWidgetVisualizer):
+
+2. In the constructor call, call the base class constructor with the matplotlib visualizer class
+   as ``pic_vis_cls`` keyword.
+
+   The base class will then create an instance of the visualizer class and
+   delegate the plotting job to it.
+
+.. code:: python
+
+   # taken from lib/python/picongpu/plugins/jupyter_widgets/energy_histogram_visualizer.py
+   from .base_visualizer import Visualizer as BaseWidgetVisualizer
+   from picongpu.plugins.plot_mpl.energy_histogram_visualizer import Visualizer \
+        as PICVisualizer
+
+   class Visualizer(BaseWidgetVisualizer):
+       def __init__(self, experiment_path, fig=None, your_additional_args):
+
+           BaseWidgetVisualizer.__init__(self,
+                                         experiment_path,
+                                         pic_vis_cls=PICVisualizer,
+                                         fig=fig)
+           # do stuff with your_additional_args
+
+
+3. (optional) adjust the ``_basic_ui_elements(self)`` function.
+    The base class already creates dropdown widgets for interactive selection of scan and simulation
+    directories as well as a slider to adjust the simulation time (i.e. the iteration).
+    If your class does not need some of those widgets, feel free to override this function.
+    It needs to return a list of widgets.
+    Note that scan and simulation dropdowns should probably be returned in any case to allow switching
+    between different simulations.
+
+4. implement the  ``_extra_ui_elements(self)`` function.
+
+    In this function jupyter widgets have to be defined as member variables of the class
+    to allow interactive manipulation of parameters the underlying
+    matplotlib visualizer is capable of handling.
+    It needs to return a list of those widgets.
+
+.. code:: python
+
+   # taken from lib/python/picongpu/plugins/jupyter_widgets/energy_histogram_visualizer.py
+   def _extra_ui_elements(self):
+       """
+       Create and return as a list the individual widgets that are necessary
+       for this special visualizer.
+       """
+       self.species = widgets.Dropdown(description="Species",
+                                       options=["e"],
```

Comment:
`species` and `species_filter` options will need to be sett-able or auto-readible from a `run_directory`. They are variables of each individual simulation